packages/backend-app-api/src/services/implementations/discovery/MultipleBackendHostDiscovery.ts

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { HostDiscovery, TokenManager } from '@backstage/backend-common';
+import {
+  DiscoveryService,
+  LoggerService,
+  RootFeatureRegistryService,
+} from '@backstage/backend-plugin-api';
+import { Config } from '@backstage/config';
+import fetch from 'node-fetch';
+import { NotFoundError } from '@backstage/errors';
+import Cache from 'node-cache';
+
+/** @public */
+export interface DiscoveryUrl {
+  internal: string;
+  external: string;
+}
+
+/** @public */
+export type PluginRegistrations = Record<string, DiscoveryUrl>;
+
+/** @public */
+export class MultipleBackendHostDiscovery implements DiscoveryService {
+  #gatewayUrl: string;
+  #instanceUrl: string;
+
+  #isGateway: boolean;
+  #isInitialized = false;
+
+  #discovery: DiscoveryService;
+  #rootFeatureRegistry: RootFeatureRegistryService;
+  #tokenManager: TokenManager;
+  #logger: LoggerService;
+
+  /** A cache for non-gateway instances to store gateway routings for a set period of time. */
+  #gatewayResponseCache: Cache;
+
+  #instanceCheckInCache: Cache;
+
+  // A map of plugin to URLs.
+  #plugins: PluginRegistrations = {};
+  // A map of instance URL to plugin.
+  #instancePlugins: Record<string, Set<string>> = {};
```

Comment:
We'll need this to be backed by database or similar right? It must be possible to scale the gateway horizontally and allow any gateway node to receive registrations/queries.

Reply:
Yup... my brain was thinking of just a single gateway node but that wouldn't be all that useful.